Support for RLS in base tables

thoshiai opened the pull request:
When IVM execute calculation of incremental difference, it use Ephemeral Named Relation(ENR) for store of base table diff. VIM was not supported RLS, because ENR could not have RLS info.

It meant that data which owner don't have permission was seen on matview. So this patch add that give RLS info of base table to RTE of ENR.

thoshiai committed:
Support for RLS in base tables

When IVM execute calculation of incremental difference, it use Ephemeral Named Relation(ENR) for store of base table diff.
Because ENR could nott have RLS, IVM was not supported RLS.
It meant that matview can see data which owner don't have permission.
this patch add that give RLS of base table to RTE of enr.

yugo-n commented:
Thanks!  BTW, could you also add the test for confirming if it works as you intended?

thoshiai committed:
Add regression test for IVM with row level security

thoshiai commented:
Thank you for your comment. I added test case.

thoshiai commented:
> 
> 
> Thanks! BTW, could you also add the test for confirming if it works as you intended?

Thank you for your comment. I added test case.

yugo-n commented:
@thoshiai Could you please resolve the conflicts?

yugo-n commented on src/backend/commands/matview.c:
```diff
@@ -4256,3 +4267,45 @@ isIvmColumn(const char *s)
+	get_row_security_policies(query, rte, 1,
+							  &securityQuals, &withCheckOptions,
+							  &hasRowSecurity, &hasSubLinks);
+	rte->securityQuals = list_concat(securityQuals, rte->securityQuals);
```
Why is this list_concat necessary? It seems that this RTE is not used at all, although I might be missing something, though.

yugo-n replied:
> > @thoshiai Could you please resolve the conflicts?
> 
> I resolved conflict

Thanks!

yugo-n replied:
>Why is this list_concat necessary? It seems that this RTE is not used at all, although I might be missing something, though.

After some investigation, I think this list_concat is unnecessary.  Maybe you borrowed this code from fireRIRrules() which needs to rewrite RTE, but get_securityQuals() just needs to get the security guals of a base table for ENR ,and RTE of a base table is not necessary.  Actually, the regression test works well even if we comments this list_concat out.

thoshiai replied:
OK, I think so, too.

yugo-n replied:
@thoshiai list_concat is still left in the latest repository. Can we remove this?

thoshiai replied:
Sorry , I missed it.

thoshiai committed:
Merge branch 'IVM' into IVM_rls

resolve coflict with latest IVM branch

Conflicts:
	src/backend/commands/matview.c
	src/test/regress/expected/incremental_matview.out
	src/test/regress/sql/incremental_matview.sql

thoshiai committed:
resolve merge conflict

This has regression test with cassert is all passed.

thoshiai commented:
> 
> 
> @thoshiai Could you please resolve the conflicts?

I resolved conflict

yugo-n commented on src/backend/commands/matview.c:
```diff
@@ -284,7 +285,7 @@ static void mv_HashPreparedPlan(MV_QueryKey *key, SPIPlanPtr plan);
 static void clean_up_IVM_hash_entry(MV_TriggerHashEntry *entry);
 static void clean_up_IVM_temptable(Oid tempOid_old, Oid tempOid_new);
-
+static List *get_securityQuals(Oid relId,Query *query);
```
Very very little point, but I want a space from the comma.

thoshiai replied:
Sorry, I will fix it.

yugo-n commented on src/backend/commands/matview.c:
```diff
@@ -4338,3 +4349,47 @@ isIvmColumn(const char *s)
+ * This is used by IVM for copying RLS from base table to enr.
+ */
+static List *
+get_securityQuals(Oid relId,Query *query)
```
I want a space, again. ;-)

thoshiai replied:
Sorry, I will fix it too...

yugo-n commented on src/backend/commands/matview.c:
```diff
@@ -1754,6 +1755,9 @@ register_delta_ENRs(ParseState *pstate, Query *query, List *tables)
 			nsitem = addRangeTableEntryForENR(pstate, makeRangeVar(NULL, enr->md.name, -1), true);
 			rte = nsitem->p_rte;
+			/* if base table has RLS, set secuirty condition to enr*/
+			rte->securityQuals = get_securityQuals(table->table_id, query);
```
Maybe get_securityQuals is not necessary in register_delta_ENRs at least for the current implementation because ENRs are not directly accessed in the primary query rather these are used in subqueries built from SQL strings. However, it will harmless and might be safety.

thoshiai replied:
Thank you for your comment.
I think, as safety is important, this is fine.

yugo-n commented on src/backend/commands/matview.c:
```diff
@@ -4338,3 +4349,47 @@ isIvmColumn(const char *s)
+	nsitem = addRangeTableEntryForRelation(pstate, rel, AccessShareLock, NULL, false, false);
+	rte = nsitem->p_rte;
+
+	get_row_security_policies(query, rte, 1,
```
I would like an explanation for why the third argument is 1. Also, I think some comments is necessary for the explanation.  In my understanding, this is because all ENRs are accessed in a subquery which has only one RTE and rtindex is always 1, right?

thoshiai replied:
What you say is correct.
Currently, get_securityQuals function is used by IVM only if rt_index is 1.
Althouth it is no problem to set rt_index to 1, it is dengerous to use fixed value. I would like to add new argument rt_index on get_securityQuals().

yugo-n commented:
I found this works wrong when multiple tables are simultaneously updated.
```sql
CREATE TABLE rls_tbl(id int, data text, owner name);
CREATE TABLE
INSERT INTO rls_tbl VALUES
  (1,'foo','rls_user'),
  (2,'bar','postgres');
INSERT 0 2
CREATE POLICY rls_tbl_policy ON rls_tbl FOR SELECT TO PUBLIC USING(owner = current_user);
CREATE POLICY
ALTER TABLE rls_tbl ENABLE ROW LEVEL SECURITY;
ALTER TABLE
GRANT ALL on rls_tbl TO PUBLIC;
GRANT
CREATE TABLE num_tbl(i int, num text);
CREATE TABLE
INSERT INTO num_tbl VALUES 
 (1,'one'),(2,'two'),(3,'three'),(4,'four');
INSERT 0 4
GRANT ALL on num_tbl TO PUBLIC;
GRANT
SET SESSION AUTHORIZATION rls_user;
SET
CREATE INCREMENTAL MATERIALIZED VIEW  immv_rls AS SELECT * FROM rls_tbl JOIN num_tbl ON i=id;
SELECT 1
RESET SESSION AUTHORIZATION;
RESET
SELECT * FROM immv_rls;
 id | data |  owner   | i | num 
----+------+----------+---+-----
  1 | foo  | rls_user | 1 | one
(1 row)

INSERT INTO rls_tbl VALUES
  (3,'baz','rls_user'),
  (4,'qux','postgres');
INSERT 0 2
SELECT * FROM immv_rls;
 id | data |  owner   | i |  num  
----+------+----------+---+-------
  1 | foo  | rls_user | 1 | one
  3 | baz  | rls_user | 3 | three
(2 rows)

WITH 
 x AS (UPDATE rls_tbl SET data = data || '_2' where id in (3,4)),
 y AS (UPDATE num_tbl SET num = num || '_2' where i in (3,4))
SELECT;
--
(1 row)

SELECT * FROM immv_rls;
 id | data  |  owner   | i |   num   
----+-------+----------+---+---------
  1 | foo   | rls_user | 1 | one
  4 | qux   | postgres | 4 | four_2
  3 | baz_2 | rls_user | 3 | three_2
(3 rows)
```

Note that the last result contains a tuple whose owner is not rls_user although this must not be invisible for the matview owner.  I think this is due to ENRs in a subquery to calculate the pre-state, so we need to set securtyQuals for them in get_prestate_rte.

Attached is the test SQL script.

[test_rls.sql.txt](https://github.com/sraoss/pgsql-ivm/files/4162789/test_rls.sql.txt)

thoshiai commented:
> I found this works wrong when multiple tables are simultaneously updated.

Thank you for your comment and sharing test code.
I  will confirm and fix it.

thoshiai committed:
Merge branch 'IVM' into IVM_rls

thoshiai committed:
Fix some mistakes and problem for review comment

Add new argument which is rt_index in get_securityQuals() function,
because rt_index is able to be not found in rte.
And get_prestate_rte() call get_securityQuals() too and it is executed
by relowner, bacause it use tuplestores too.

thoshiai commented:
> I found this works wrong when multiple tables are simultaneously updated.

I check replaying this problem. And I add a commit that it is resolved and add regrresion test.

yugo-n commented on src/backend/commands/matview.c:
```diff
@@ -1843,6 +1850,20 @@ get_prestate_rte(RangeTblEntry *rte, MV_TriggerTable *table,
+	foreach(lc, sub->rtable)
+	{
+		RangeTblEntry *rte = (RangeTblEntry *)lfirst(lc);
+		/* if contain UNION ALL clauses, rte has a subquey */
```
You means:
"If this query contains UNION ALL clauses, RTEs in rtables has a subquery which contains ENR" ? 

I think that it is straight forward to check setOperations is NIL or not. If this is not NIL, RTEs must have a subquery and this can be confirmed by Assert.

thoshiai replied:
> You means:
> "If this query contains UNION ALL clauses, RTEs in rtables has a subquery which contains ENR" ?

That's right.

> I think that it is straight forward to check setOperations is NIL or not. If this is not NIL, RTEs must have a subquery and this can be confirmed by Assert.

Thank you for your alternative suggestion! I use setOperations for check.

thoshiai committed:
Fix a mistake and problem for review comments (take2)

yugo-n commented:
Thanks! It seems good.
